// @storybook/ui/dist/app.js
"use strict";

require("core-js/modules/es.object.assign");

require("core-js/modules/es.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _theming = require("@storybook/theming");

var _reactResizeDetector = _interopRequireDefault(require("react-resize-detector"));

var _memoizerific = _interopRequireDefault(require("memoizerific"));

var _router = require("@storybook/router");

var _mobile = require("./components/layout/mobile");

var _desktop = require("./components/layout/desktop");

var _nav = _interopRequireDefault(require("./containers/nav"));

var _preview = _interopRequireDefault(require("./containers/preview"));

var _panel = _interopRequireDefault(require("./containers/panel"));

var _notifications = _interopRequireDefault(require("./containers/notifications"));

var _settings = _interopRequireDefault(require("./settings"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _ref =
/*#__PURE__*/
_react["default"].createElement(_settings["default"], null);

var createProps = (0, _memoizerific["default"])(1)(function () {
  return {
    Nav: _nav["default"],
    Preview: _preview["default"],
    Panel: _panel["default"],
    Notifications: _notifications["default"],
    pages: [{
      key: 'settings',
      render: function render() {
        return _ref;
      },
      // eslint-disable-next-line react/prop-types
      route: function route(_ref2) {
        var children = _ref2.children;
        return _react["default"].createElement(_router.Route, {
          path: "/settings",
          startsWith: true
        }, children);
      }
    }]
  };
});

var View = _theming.styled.div({
  position: 'fixed',
  overflow: 'hidden',
  height: '100vh',
  width: '100vw'
});

var _ref4 =
/*#__PURE__*/
_react["default"].createElement(_theming.Global, {
  styles: _theming.createGlobal
});

var _ref6 =
/*#__PURE__*/
_react["default"].createElement("div", null);

var App = _react["default"].memo(function (_ref3) {
  var viewMode = _ref3.viewMode,
      layout = _ref3.layout;
  var props = createProps();
  return _react["default"].createElement(View, null, _ref4, _react["default"].createElement(_reactResizeDetector["default"], {
    handleWidth: true,
    handleHeight: true
  }, function (_ref5) {
    var width = _ref5.width,
        height = _ref5.height;

    if (!width || !height) {
      return _ref6;
    }

    if (width < 600) {
      return _react["default"].createElement(_mobile.Mobile, _extends({}, props, {
        viewMode: viewMode,
        options: layout
      }));
    }

    return _react["default"].createElement(_desktop.Desktop, _extends({}, props, {
      viewMode: viewMode,
      options: layout
    }, {
      width: width,
      height: height
    }));
  }));
});

App.propTypes = {
  viewMode: _propTypes["default"].oneOf(['story', 'info']),
  layout: _propTypes["default"].shape({}).isRequired
};
App.defaultProps = {
  viewMode: undefined
};
App.displayName = 'App';
var _default = App;
exports["default"] = _default;
