// @storybook/ui/dist/components/sidebar/Sidebar.stories.js
"use strict";

require("core-js/modules/es.object.define-property");

require("core-js/modules/es.object.get-own-property-descriptor");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loading = exports.simple = exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _Sidebar = _interopRequireDefault(require("./Sidebar"));

var SidebarHeadingStories = _interopRequireWildcard(require("./SidebarHeading.stories"));

var SidebarStoriesStories = _interopRequireWildcard(require("./SidebarStories.stories"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  Component: _Sidebar["default"],
  title: 'UI|Sidebar/Sidebar'
};
exports["default"] = _default;
var menu = SidebarHeadingStories.standard.storyData.menu;
var _SidebarStoriesStorie = SidebarStoriesStories.withRoot.storyData,
    stories = _SidebarStoriesStorie.stories,
    storyId = _SidebarStoriesStorie.storyId;

var _ref =
/*#__PURE__*/
_react["default"].createElement(_Sidebar["default"], {
  menu: menu,
  stories: stories,
  storyId: storyId
});

var simple = function simple() {
  return _ref;
};

exports.simple = simple;
simple.displayName = "simple";
simple.storyData = {
  menu: menu,
  stories: stories,
  storyId: storyId
};

var loading = function loading() {
  return _react["default"].createElement(_Sidebar["default"], {
    menu: menu,
    stories: {},
    loading: true
  });
};

exports.loading = loading;
loading.displayName = "loading";
loading.storyData = {
  menu: menu,
  stories: {}
};
