// @storybook/ui/dist/components/notifications/item.stories.js
"use strict";

require("core-js/modules/es.object.assign");

require("core-js/modules/es.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withLink = exports.longText = exports.simple = exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _item = _interopRequireDefault(require("./item"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  component: _item["default"],
  title: 'UI|Notifications/Item',
  decorators: [function (storyFn) {
    return _react["default"].createElement("div", {
      style: {
        width: '240px',
        margin: '1rem'
      }
    }, storyFn());
  }]
};
exports["default"] = _default;

function buildExample(notification) {
  var _ref =
  /*#__PURE__*/
  _react["default"].createElement(_item["default"], {
    notification: notification
  });

  var story = function story() {
    return _ref;
  };

  story.storyData = {
    notification: notification
  };
  return story;
}

var simple = buildExample({
  content: '🎉 Storybook is cool!'
});
exports.simple = simple;
var longText = buildExample({
  content: '🎉 This is a long message that extends over two lines!'
});
exports.longText = longText;
var withLink = buildExample(Object.assign({}, simple.storyData.notification, {
  content: '🎉 Storybook X.X is available! Download now »',
  link: '/some/path'
}));
exports.withLink = withLink;
