// @storybook/ui/dist/components/notifications/notifications.stories.js
"use strict";

require("core-js/modules/es.array.filter");

require("core-js/modules/es.array.map");

require("core-js/modules/es.array.slice");

require("core-js/modules/es.date.to-string");

require("core-js/modules/es.object.assign");

require("core-js/modules/es.object.define-property");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.object.values");

require("core-js/modules/es.regexp.to-string");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.placement = exports.all = exports.single = exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _notifications = _interopRequireDefault(require("./notifications"));

var NotificationItemStories = _interopRequireWildcard(require("./item.stories"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  component: _notifications["default"],
  title: 'UI|Notifications/Notifications',
  decorators: [function (storyFn) {
    return _react["default"].createElement("div", {
      style: {
        width: '240px',
        margin: '1rem'
      }
    }, storyFn());
  }]
};
exports["default"] = _default;
var notifications = Object.values(NotificationItemStories).filter(function (s) {
  return s.storyData;
}).map(function (_ref, index) {
  var notification = _ref.storyData.notification;
  return Object.assign({}, notification, {
    id: index.toString()
  });
});

var single = function single() {
  return _react["default"].createElement(_notifications["default"], {
    notifications: notifications.slice(0, 1)
  });
};

exports.single = single;
single.displayName = "single";
single.storyData = {
  notifications: notifications.slice(0, 1)
};

var _ref2 =
/*#__PURE__*/
_react["default"].createElement(_notifications["default"], {
  notifications: notifications
});

var all = function all() {
  return _ref2;
};

exports.all = all;
all.displayName = "all";
all.storyData = {
  notifications: notifications
};

var placement = function placement() {
  return _react["default"].createElement(_notifications["default"], {
    placement: {
      position: 'fixed',
      left: 20,
      bottom: 20
    },
    notifications: notifications
  });
};

exports.placement = placement;
placement.displayName = "placement";
all.storyData = {
  notifications: notifications
};
