// @storybook/ui/dist/components/preview/background.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.object.create");

require("core-js/modules/es.object.define-property");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.get-prototype-of");

require("core-js/modules/es.object.set-prototype-of");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BackgroundProvider = exports.BackgroundConsumer = exports.Background = exports.Grid = void 0;

var _react = _interopRequireWildcard(require("react"));

var _server = require("react-dom/server");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _theming = require("@storybook/theming");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var Context = _react["default"].createContext();

var Provider =
/*#__PURE__*/
function (_Component) {
  _inherits(Provider, _Component);

  function Provider() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Provider);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Provider)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      grid: false,
      value: 'transparent'
    };

    _this.setValue = function (value) {
      return _this.setState({
        value: value
      });
    };

    _this.setGrid = function (grid) {
      return _this.setState({
        grid: grid
      });
    };

    return _this;
  }

  _createClass(Provider, [{
    key: "render",
    value: function render() {
      var setValue = this.setValue,
          setGrid = this.setGrid;
      var children = this.props.children;
      var _this$state = this.state,
          value = _this$state.value,
          grid = _this$state.grid;
      return _react["default"].createElement(Context.Provider, {
        value: {
          value: value,
          setValue: setValue,
          grid: grid,
          setGrid: setGrid
        }
      }, children);
    }
  }]);

  return Provider;
}(_react.Component);

exports.BackgroundProvider = Provider;
Provider.displayName = "Provider";
Provider.propTypes = {
  children: _propTypes["default"].node.isRequired
};
var Consumer = Context.Consumer;
exports.BackgroundConsumer = Consumer;

var _ref =
/*#__PURE__*/
_react["default"].createElement("rect", {
  width: "100%",
  height: "100%",
  fill: "url(#grid)"
});

function createGridStyles(cellSize) {
  var cellSizeDoubled = cellSize * 2;
  var cellSizeSquared = Math.pow(cellSize, 2);
  var gridSVGEncoded = encodeURIComponent((0, _server.renderToStaticMarkup)(_react["default"].createElement("svg", {
    width: "100%",
    height: "100%",
    xmlns: "http://www.w3.org/2000/svg"
  }, _react["default"].createElement("defs", null, _react["default"].createElement("pattern", {
    id: "smallGrid",
    width: cellSize,
    height: cellSize,
    patternUnits: "userSpaceOnUse"
  }, _react["default"].createElement("path", {
    d: "M ".concat(cellSize, " 0 L 0 0 0 ").concat(cellSize),
    fill: "none",
    stroke: "gray",
    strokeWidth: "0.5"
  })), _react["default"].createElement("pattern", {
    id: "grid",
    width: cellSizeSquared,
    height: cellSizeSquared,
    patternUnits: "userSpaceOnUse"
  }, _react["default"].createElement("rect", {
    width: cellSizeSquared,
    height: cellSizeSquared,
    fill: "url(#smallGrid)"
  }), _react["default"].createElement("path", {
    d: "M ".concat(cellSizeSquared, " 0 L 0 0 0 ").concat(cellSizeSquared),
    fill: "none",
    stroke: "gray",
    strokeWidth: "1"
  }))), _ref)));
  return {
    backgroundImage: "url(\"data:image/svg+xml,".concat(gridSVGEncoded, "\")"),
    backgroundSize: "".concat(cellSizeSquared, "px ").concat(cellSizeSquared, "px, ").concat(cellSizeSquared, "px ").concat(cellSizeSquared, "px, ").concat(cellSizeDoubled, "px ").concat(cellSizeDoubled, "px, ").concat(cellSizeDoubled, "px ").concat(cellSizeDoubled, "px"),
    backgroundPosition: '-2px -2px',
    mixBlendMode: 'difference'
  };
}

var Grid = _theming.styled.div({
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%'
}, function (_ref2) {
  var theme = _ref2.theme;
  return createGridStyles(theme.background.gridCellSize);
});

exports.Grid = Grid;

var Background = _theming.styled.div({
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%',
  transition: 'background .1s linear',
  iframe: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0,
    left: 0,
    border: '0 none'
  }
}, function (_ref3) {
  var theme = _ref3.theme;
  return {
    background: theme.background.content
  };
});

exports.Background = Background;
