// @storybook/core-events/dist/index.js
"use strict";

require("core-js/modules/es.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STORY_THREW_EXCEPTION = exports.STORY_CHANGED = exports.STORY_ERRORED = exports.STORY_MISSING = exports.STORY_RENDERED = exports.STORY_RENDER = exports.STORY_ADDED = exports.STORY_INIT = exports.REGISTER_SUBSCRIPTION = exports.FORCE_RE_RENDER = exports.PREVIEW_KEYDOWN = exports.SELECT_STORY = exports.STORIES_CONFIGURED = exports.SET_STORIES = exports.GET_STORIES = exports.SET_CURRENT_STORY = exports.GET_CURRENT_STORY = exports.CHANNEL_CREATED = exports["default"] = void 0;
var events; // Enables: `import Events from ...`

(function (events) {
  events["CHANNEL_CREATED"] = "channelCreated";
  events["GET_CURRENT_STORY"] = "getCurrentStory";
  events["SET_CURRENT_STORY"] = "setCurrentStory";
  events["GET_STORIES"] = "getStories";
  events["SET_STORIES"] = "setStories";
  events["STORIES_CONFIGURED"] = "storiesConfigured";
  events["SELECT_STORY"] = "selectStory";
  events["PREVIEW_KEYDOWN"] = "previewKeydown";
  events["STORY_ADDED"] = "storyAdded";
  events["STORY_CHANGED"] = "storyChanged";
  events["STORY_UNCHANGED"] = "storyUnchanged";
  events["FORCE_RE_RENDER"] = "forceReRender";
  events["REGISTER_SUBSCRIPTION"] = "registerSubscription";
  events["STORY_INIT"] = "storyInit";
  events["STORY_RENDER"] = "storyRender";
  events["STORY_RENDERED"] = "storyRendered";
  events["STORY_MISSING"] = "storyMissing";
  events["STORY_ERRORED"] = "storyErrored";
  events["STORY_THREW_EXCEPTION"] = "storyThrewException";
})(events || (events = {}));

var _default = events; // Enables: `import * as Events from ...` or `import { CHANNEL_CREATED } as Events from ...`
// This is the preferred method

exports["default"] = _default;
var _events = events,
    CHANNEL_CREATED = _events.CHANNEL_CREATED;
exports.CHANNEL_CREATED = CHANNEL_CREATED;
var _events2 = events,
    GET_CURRENT_STORY = _events2.GET_CURRENT_STORY;
exports.GET_CURRENT_STORY = GET_CURRENT_STORY;
var _events3 = events,
    SET_CURRENT_STORY = _events3.SET_CURRENT_STORY;
exports.SET_CURRENT_STORY = SET_CURRENT_STORY;
var _events4 = events,
    GET_STORIES = _events4.GET_STORIES;
exports.GET_STORIES = GET_STORIES;
var _events5 = events,
    SET_STORIES = _events5.SET_STORIES;
exports.SET_STORIES = SET_STORIES;
var _events6 = events,
    STORIES_CONFIGURED = _events6.STORIES_CONFIGURED;
exports.STORIES_CONFIGURED = STORIES_CONFIGURED;
var _events7 = events,
    SELECT_STORY = _events7.SELECT_STORY;
exports.SELECT_STORY = SELECT_STORY;
var _events8 = events,
    PREVIEW_KEYDOWN = _events8.PREVIEW_KEYDOWN;
exports.PREVIEW_KEYDOWN = PREVIEW_KEYDOWN;
var _events9 = events,
    FORCE_RE_RENDER = _events9.FORCE_RE_RENDER;
exports.FORCE_RE_RENDER = FORCE_RE_RENDER;
var _events10 = events,
    REGISTER_SUBSCRIPTION = _events10.REGISTER_SUBSCRIPTION;
exports.REGISTER_SUBSCRIPTION = REGISTER_SUBSCRIPTION;
var _events11 = events,
    STORY_INIT = _events11.STORY_INIT;
exports.STORY_INIT = STORY_INIT;
var _events12 = events,
    STORY_ADDED = _events12.STORY_ADDED;
exports.STORY_ADDED = STORY_ADDED;
var _events13 = events,
    STORY_RENDER = _events13.STORY_RENDER;
exports.STORY_RENDER = STORY_RENDER;
var _events14 = events,
    STORY_RENDERED = _events14.STORY_RENDERED;
exports.STORY_RENDERED = STORY_RENDERED;
var _events15 = events,
    STORY_MISSING = _events15.STORY_MISSING;
exports.STORY_MISSING = STORY_MISSING;
var _events16 = events,
    STORY_ERRORED = _events16.STORY_ERRORED;
exports.STORY_ERRORED = STORY_ERRORED;
var _events17 = events,
    STORY_CHANGED = _events17.STORY_CHANGED;
exports.STORY_CHANGED = STORY_CHANGED;
var _events18 = events,
    STORY_THREW_EXCEPTION = _events18.STORY_THREW_EXCEPTION;
exports.STORY_THREW_EXCEPTION = STORY_THREW_EXCEPTION;
